Sync/Synchronizers/Bookmarks/BookmarksSynchronizer.swift

```diff
@@ -89,6 +89,49 @@ open class BufferingBookmarksSynchronizer: TimestampedSingleCollectionSynchroniz
+        let newBookmarkGUIDs = additionalChildren.map { $0.guid }
+        return buffer.getBufferItemWithGUID(BookmarkRoots.MobileFolderGUID).bind { maybeMobileRoot in
+            // Update (or create!) the Mobile Root folder with its new children.
+            if let mobileRoot = maybeMobileRoot.successValue {
+                return buffer.getBufferChildrenGUIDsForParent(mobileRoot.guid)
+                    .map { $0.map({ (mobileRoot: mobileRoot, children: $0 + newBookmarkGUIDs) }) }
+            } else {
+                return local.getLocalItemWithGUID(BookmarkRoots.MobileFolderGUID)
+                    .map { $0.map({ (mobileRoot: $0, children: newBookmarkGUIDs) }) }
+            }
+        } >>== { (mobileRoot: BookmarkMirrorItem, children: [GUID]) in
+            let payload = mobileRoot.asPayloadWithChildren(children)
+            let mappedGUID = payload["id"].string ?? mobileRoot.guid
+            return deferMaybe(Record<BookmarkBasePayload>(id: mappedGUID, payload: payload))
+        }
+    }
+
+    func buildMobileRootAndChildrenRecords(_ local: LocalItemSource, _ buffer: BufferItemSource, additionalChildren: [BookmarkMirrorItem]) -> Deferred<Maybe<(mobileRootRecord: Record<BookmarkBasePayload>, childrenRecords: [Record<BookmarkBasePayload>])>> {
+        let childrenRecords = additionalChildren.map { bkm -> Record<BookmarkBasePayload> in
+            let payload = bkm.asPayload()
+            let mappedGUID = payload["id"].string ?? bkm.guid
+            return Record<BookmarkBasePayload>(id: mappedGUID, payload: payload)
+        }
+
+        return self.buildMobileRootRecord(local, buffer, additionalChildren: additionalChildren) >>== { mobileRootRecord in
+            return deferMaybe((mobileRootRecord: mobileRootRecord, childrenRecords: childrenRecords))
+        }
+    }
+
+    func uploadSomeLocalRecords(_ storage: SyncableBookmarks & LocalItemSource & MirrorItemSource, _ buffer: BufferItemSource, _ mirrorer: BookmarksMirrorer, _ bookmarksClient: Sync15CollectionClient<BookmarkBasePayload>, mobileRootRecord: Record<BookmarkBasePayload>, childrenRecords: [Record<BookmarkBasePayload>]) -> Success {
+        let records = [mobileRootRecord] + childrenRecords
+        return self.uploadRecordsSingleBatch(records, lastTimestamp: mirrorer.lastModified, storageClient: bookmarksClient) >>== { (timestamp: Timestamp, succeeded: [GUID]) -> Success in
+            mirrorer.advanceNextDownloadTimestampTo(timestamp: timestamp) // We need to advance our batching downloader timestamp to match. See Bug 1253458.
+
+            let newBookmarkGUIDs = childrenRecords.map { BookmarkRoots.translateIncomingRootGUID($0.id) }
+            let bufferValuesToMoveFromLocal = Set(newBookmarkGUIDs).intersection(Set(succeeded))
+            let mobileRoot = (mobileRootRecord.payload as MirrorItemable).toMirrorItem(timestamp)
```

Comment:
We do need an updated children field, which was why I introduced the clone method earlier in my PR. I think this is fine though.